storage: avoid client-side retries on CPut/Inc write-too-old errors

spencerkimball opened the pull request:
Previously, a conditional put or increment ops that experienced a write-too-old error
would not immediately return an error. Instead, it would proceed to lay
down the intent, set the `WriteTooOld` bool on the `Transaction`, and
continue so that intents might be laid down to assure succes on the
eventual retry. The `EndTransaction` request triggers restart at commit
time in the event that `WriteTooOld` is set.
    
The problem with this is that refreshing the conditional put key is
guaranteed to fail because there is a known, newer version of the key
written. This implies a client-side retry.
    
This change makes conditional puts immediately return a `WriteTooOldError`
instead of just swallowing the error and setting the `WriteTooOld`
flag. The returned error allows the `TxnCoordSender` to refresh all
spans mid-transaction, and hopefully retry the conditional put at the
advanced timestamp, to either succeed or fail on the conditional put.
This allows the transaction to continue on success, or return a
`ConditionFailedError` on failure.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/22234)
<!-- Reviewable:end -->

spencerkimball commented:
@nvanbenschoten, ignore first commit; that's the serializable txn retries. This is the change I was telling you about to avoid client-side transaction retries for conditional puts.

@andreimatei this solves the case where the old auto-retry mechanism might have saved us from a client visible restart with conditional puts that we whiteboarded yesterday.

nvanbenschoten commented:
The code all looks good. I do have a general question about `WriteTooOldErrors` - now that we have the ability to refresh spans within a txn to a higher timestamp, what is the benefit of ever setting the `WriteTooOld` flag on a transaction and then continuing to lay down intents at the older timestamp? Why don't we always do what we're doing here and let `WriteTooOldErrors` fall back to the `TxnCoordSender`?

---

Review status: 0 of 43 files reviewed at latest revision, all discussions resolved, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/22234#-:-L4C0wzpZC2Y2axJAMKK:b-93yjwf)*
<!-- Sent from Reviewable.io -->

spencerkimball commented:
Because you end up with `O(n^2)` behavior on big transactions that encounter a lot of `WriteTooOldError`s on `Put`s. Each one has to refresh all of the spans which came before. Laying down the intent and refreshing at the end is the better algorithm.

---

Review status: 0 of 43 files reviewed at latest revision, all discussions resolved, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/22234#-:-L4D2PZWC5J8xnVQV40V:b-2zueqe)*
<!-- Sent from Reviewable.io -->

spencerkimball committed:
storage: avoid client-side retries on CPut write-too-old errors

Previously, a conditional put op that experienced a write-too-old error
would not immediately return an error. Instead, it would proceed to lay
down the intent, set the `WriteTooOld` bool on the `Transaction`, and
continue so that intents might be laid down to assure succes on the
eventual retry. The `EndTransaction` request triggers restart at commit
time in the event that `WriteTooOld` is set.

The problem with this is that refreshing the conditional put key is
guaranteed to fail because there is a known, newer version of the key
written. This implies a client-side retry.

This change makes conditional puts immediately return a `WriteTooOldError`
instead of just swallowing the error and setting the `WriteTooOld`
flag. The returned error allows the `TxnCoordSender` to refresh all
spans mid-transaction, and hopefully retry the conditional put at the
advanced timestamp, to either succeed or fail on the conditional put.
This allows the transaction to continue on success, or return a
`ConditionFailedError` on failure.

Release note: None

spencerkimball commented:
OK, this is ready. I updated it to also include `Increment` requests, which fall into the same category.

bdarnell commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Review status: 0 of 2 files reviewed at latest revision, all discussions resolved, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/22234#-:-L4IN2irdfNyajoHygi_:bnfp4nl)*
<!-- Sent from Reviewable.io -->

nvanbenschoten commented:
yep <img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>, thought I gave that here already but I guess not.

---

Review status: 0 of 2 files reviewed at latest revision, all discussions resolved, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/22234#-:-L4IOSXevGcQuDsiTXol:bg9wswh)*
<!-- Sent from Reviewable.io -->

spencerkimball commented:
Added fixes for a test in `pkg/sql/` which had become flaky from the original serializable transaction improvements change.

andreimatei commented:
Spencer what was the test fix?

---

Review status: 0 of 3 files reviewed at latest revision, all discussions resolved, all commit checks successful.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/22234#-:-L4MZd7XSheuVGqDQ1gR:b-iw77lb)*
<!-- Sent from Reviewable.io -->
